src/main/java/com/bobcat00/viaversionstatus/Listeners.java

```diff
@@ -234,81 +234,116 @@ public void onPlayerJoin(PlayerJoinEvent e)
-                plugin.getLogger().info("Executing command " + warnCommand);
-                try
-                {
-                    Bukkit.getServer().dispatchCommand(Bukkit.getConsoleSender(), warnCommand);
-                }
-                catch (CommandException exc)
-                {
-                    plugin.getLogger().info("Command returned exception: " + exc.getMessage());
+            }
+        }
+
+        // 4. Warn player if they are connecting using a newer version
+
+        if (plugin.config.getNewerVersionWarnPlayers() &&
+            !serverVersion.equals("UNKNOWN") &&
+            (clientProtocol.getId() > serverProtocol.getId()) &&
+            !player.hasPermission("viaversionstatus.exempt.warn.newer"))
+        {
+            if (!player.hasPermission("viaversionstatus.exempt.warn.newer.message")) {
+                // Delay by 250 msec (5 ticks) to make sure the player sees the message
+                Bukkit.getScheduler().runTaskLater(plugin, new Runnable() {
+                    @Override
+                    public void run() {
+                        if (player.isOnline()) {
+                            player.sendMessage(ChatColor.translateAlternateColorCodes('&',
+                                    plugin.config.getNewerVersionWarnString().replace("%player%", player.getName()).
+                                                                              replace("%displayname%", player.getDisplayName()).
+                                                                              replace("%version%", clientVersion).
+                                                                              replace("%server%", serverVersion)));
+                        }
+                    }
+                }, 5L); // time delay (ticks)
+            }
+
+            if (!player.hasPermission("viaversionstatus.exempt.warn.newer.command")) {
+                String newerVersionWarnCommand = plugin.config.getNewerVersionWarnCommand();
+                if (!newerVersionWarnCommand.isEmpty()) {
+                    newerVersionWarnCommand = newerVersionWarnCommand.replace("%player%", player.getName()).
+                                                                      replace("%displayname%", player.getDisplayName()).
+                                                                      replace("%version%", clientVersion).
+                                                                      replace("%server%", serverVersion);
+                    plugin.getLogger().info("Executing command " + newerVersionWarnCommand);
+                    try {
+                        Bukkit.getServer().dispatchCommand(Bukkit.getConsoleSender(), newerVersionWarnCommand);
+                    } catch (CommandException exc) {
+                        plugin.getLogger().info("Command returned exception: " + exc.getMessage());
+                    }
```

Comment:
Instead of duplicating the code, can you make a private method and pass in the permissions and strings as parameters?

Reply:
Addressed in commit `9f6e66c6b2357888d2744f0e3b6d589ebae50170`.